### DISCUSSION TITLE
How to derive paths or file names from catalog datasets

### DISCUSSION THREAD

**Discussion by 4h6h** [kedro-org/kedro#1859] [Idea]
Kedro version: 0.18.2

Is there an easy way to derive the respective paths or filenames from catalog datasets?

Background:

I am using kedro to do some experiments with yolov5. Since yolov5 is a framework in itself with its own assumptions/conditions, sometimes you need the paths to the (in this case) images rather than the images themselves directly.

Example:

Yolov5 expects a file data.yaml containing the paths to the images for train/val/test. I create the content of this file in a node of one pipeline. 
To get the paths to the respective folders of the datasets, I help myself by specifying the respective catalog entries in the configuration and using the following:

``` python
def _catalog(key):
    with KedroSession.create(package_name=PACKAGE_NAME, project_path=Path.cwd()) as session:
        return session.load_context().catalog.datasets.__dict__[key]

def model_in_data(images, annotations, params):
    ...
    _path = _catalog(params['catalog']['_name_of_a_catalog_dataset_']).__dict__['_path'])
   ...
```
Of course, this is not optimal and has some pitfalls.

What I would like to have is something like:
``` python
def model_in_data(images, annotations, params):
    ...
    _path = images._path
    _images = images.load()
    ...
```
However, as I understand it, this is not possible nowadays, because `images` itself is a `dict` of `<bound method AbstractVersionedDataSet.load of <kedro.extras.datasets.pillow.image_dataset.ImageDataSet object at ...>>`.

**Comment by noklam** [comment]
@4h6h Have you looked into Hooks already? I think this is very relevant.
https://kedro.readthedocs.io/en/latest/kedro.framework.hooks.specs.DataCatalogSpecs.html#kedro.framework.hooks.specs.DataCatalogSpecs

**Comment by 4h6h** [comment]
@noklam  That was indeed a good hint, not necessarily `DataCatalogsSpecs`, but `NodeSpecs` has been helpful for me. `NodeSpecs` has a method `before_node_run` that has access to both `catalog` and the `inputs` passed to the node. In addition, this method has an optional return value that allows the values of the datasets to be manipulated. 

I have left the actual datasets as they are. I then solved my particular use-case by programmatically extending the parameters intended for the function, which can be found in `inputs`, too.

As far as I am concerned, this question has been answered.

Thanks.

**Reply by noklam** [reply]
@4h6h Would it be possible to write a custom YOLODataset?

**Reply by 4h6h** [reply]
> @4h6h Would it be possible to write a custom YOLODataset?

I have to think about this question for a while, what it would actually mean to have a YoloDataset. But I will come back to it.

**Reply by 4h6h** [reply]
Part 1/3:
In general, yolov5 takes images as input and not only classifies objects in these images, but also marks their positions with bounding boxes (rectangles).

This means that yolov5 expects images and corresponding annotations to train the algorithm. These annotations are stored in a file for each image, using a simple text format for a list of the centres, lengths and widths of the bounding boxes (rectangles) and the associated classification indices of all objects to be classified; see [Yolov5 Label Format](https://pyimagesearch.com/2022/06/20/training-the-yolov5-object-detector-on-a-custom-dataset/#:~:text=by%20the%20model.-,YOLOv5%20Label%20Format,-Since%20we%20will).

Since this annotation format is not included in the current list of available datasets, it is the first missing part to be added (`Yolov5AnnotationDataSet`):
`_load (self) -> List: `: I myself load the annotations as a simple list of annotations. Other, more sophisticated formats such as pandas DataFrame are possible.
`_save(self, data: List) -> None:`: Likewise `_load`- I use lists, other data formats are conceivable.

**Reply by 4h6h** [reply]
Part 2/3:
Yolov5 expects a model for the training, too, i.e., rather a reference to its persisted weights *.pt in the file system; see also the initial question. 

In addition, the algorithm itself generates a number of models (or weights), at least for the last and best training run. These models can be used for various purposes, to start further training runs or to make predictions from any new data (machine learning inference).

Especially for the latter use case, the model itself is needed. Therefore, it makes sense to have a dedicated (read only) DataSet for it (`Yolov5ModelDataSetReader`). Maybe like this:

```
from yolov5.models.common import DetectMultiBackend

class Yolov5ModelDataSetReader(AbstractDataSet):

    ...

    def _load(self) -> DetectMultiBackend:
        load_path = get_filepath_str(self._filepath, self._protocol)
        dnn  = self._load_args['dnn'] if 'dnn' in self._load_args else False
        fp16 = self._load_args['fp16'] if 'fp16' in self._load_args else False
        fuse = self._load_args['fuse'] if 'fuse' in self._load_args else True

        self._logger.info('Yolov5 model dataset - load arguments - dnn: {0}, fp16: {1}, fuse: {2}'.format(dnn, fp16, fuse))

        model = DetectMultiBackend(weights=load_path, dnn=dnn, fp16=fp16, fuse=fuse)
        model.float()

        return model

    def _save(self, model) -> None:
        pass
        # or raise DataSetError(f"Saving not supported for '{self.__class__.__name__}'")

    ...

```
Additional note:
Yolov5 uses the PyTorch framework. Therefore, this reader could probably be generalised for PyTorch models.

**Reply by 4h6h** [reply]
Part 3/3:
The training results are published in a dedicated folder. During the training a number of files are created: confusion_matrix.png, labels_correlogram.jpg, P_curve.png, R_curve.png, PR_curve.png, to name a few visual files.

The options used for training are stored in opt.yaml and the hyperparameters in hyp.yaml. The corresponding datasets are available in kedro for this purpose.

The metrix data of the runs are stored in a csv file results.csv. There is a DataSet in kedro for this as well.

In addition to all the others, a file is created for Tensorboard, which can contain virtually all other artefacts, i.e. scalars, images and graphs. This file is updated per training run. Tensorboard then displays the results. Tensorboard calls the mechanism behind it [EventAccumulator](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/event_processing/event_accumulator.py).

Thus, in order to get the results in this compact form, we need another DataSetReader (`TFEventAccumulatorDataSetReader`), like this:

```
from tensorboard.backend.event_processing.event_accumulator import (
    EventAccumulator, 
    DEFAULT_SIZE_GUIDANCE, 
    STORE_EVERYTHING_SIZE_GUIDANCE,
)

class TFEventAccumulatorDataSetReader(AbstractDataSet):

    ...

    def _load(self) -> EventAccumulator:
        load_path = get_filepath_str(self._filepath, self._protocol)
        guidance  = STORE_EVERYTHING_SIZE_GUIDANCE if 'guidance' in self._load_args is 'all' else DEFAULT_SIZE_GUIDANCE
        return EventAccumulator(load_path, size_guidance=guidance)

    def _save(self, ea) -> None:
        pass
        # or raise DataSetError(f"Saving not supported for '{self.__class__.__name__}'")

    ...
```
